### PR TITLE
Issue 5441: Append Lua callstack to crash dumps

### LINKED ISSUE #5441: Add Lua callstack when writing dump
To make it easier to debug crashes, we want to append the Lua callstack (if available) to the debug dump.

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

**Comment by JCash** [comment]
Screenshot 2020-12-30 at 16 47 50

**Review by britzl — DISMISSED** [review]
Great!

**Comment by britzl** [comment]
Now that I think of it we should use the same solution for this issue where we previously relied on reference to debug.traceback(): https://github.com/defold/defold/issues/3431

But an even more comprehensive solution for getting a Lua stacktrace should be the actual implementation of debug.traceback(): https://github.com/defold/defold/blob/dev/engine/lua/src/lua/ldblib.c#L321

It seems to also support coroutines and more.

**Comment by JCash** [comment]
A (now) correct printout from inside a coroutine:

Screenshot 2021-01-04 at 11 05 30

File: engine/engine/src/engine.cpp [hunk 1]
```diff
@@ -434,6 +434,42 @@ namespace dmEngine
...
+
+    static void CrashHandlerCallback(void* ctx, char* buffer, uint32_t buffersize)
+    {
+        HEngine engine = (HEngine)ctx;
+        if (engine->m_SharedScriptContext) {
+            LuaCallstackCtx ctx;
+            ctx.m_First = true;
+            ctx.m_Buffer = buffer;
+            ctx.m_BufferSize = buffersize;
+            dmScript::GetLuaTraceback(dmScript::GetLuaState(engine->m_SharedScriptContext), "Sln", GetLuaStackTraceCbk, &ctx);
```
Comment (JCash) [inline engine/engine/src/engine.cpp:469]: How to invoke the new function.

File: engine/script/src/script.cpp [hunk 2]
```diff
@@ -258,13 +258,76 @@ namespace dmScript
...
+        else {
+            *arg = 0;
+            return L;
+        }
+    }
+
+    // From https://zeux.io/2010/11/07/lua-callstack-with-c-debugger/
+    // and also
+    // https://github.com/defold/defold/blob/dev/engine/lua/src/lua/ldblib.c#L321
+    void GetLuaTraceback(lua_State* L, const char* infostring, void (*cbk)(lua_State* L, lua_Debug* entry, void* ctx), void* ctx)
```
Comment (JCash) [inline engine/script/src/script.cpp:285]: The new function.
Compare with https://github.com/defold/defold/blob/dev/engine/lua/src/lua/ldblib.c#L321

File: engine/script/src/script.cpp [hunk 3]
```diff
@@ -258,13 +258,76 @@ namespace dmScript
...
+
+        if (lua_isnumber(L, arg+2)) {
+            level = (int)lua_tointeger(L, arg+2);
+            lua_pop(L, 1); // problematic?!
+        }
+        else {
+            level = (L == L1) ? 1 : 0;  // level 0 may be this own function
+        }
+
+        if (lua_gettop(L) != arg && !lua_isstring(L, arg+1)) return;  // message is not a string
```
Comment (JCash) [inline engine/script/src/script.cpp:306]: Kept it like for consistency with original function.

File: engine/script/src/script.cpp [hunk 4]
```diff
@@ -258,13 +258,76 @@ namespace dmScript
...
+        int arg = 0;
+        lua_State* L1 = GetLuaThread(L, &arg);
+
+        lua_Debug entry;
+        int level = 0;
+        bool first = true;
+
+        if (lua_isnumber(L, arg+2)) {
+            level = (int)lua_tointeger(L, arg+2);
+            lua_pop(L, 1); // problematic?!
```
Comment (JCash) [inline engine/script/src/script.cpp:300]: I don't see it as much of a problem in the crash handler, but what about when using it as the "debug traceback" ? Does it mess something up?

**Review by britzl — APPROVED** [review]

**Pull request merged**